src/libraries/System.Collections.Concurrent/src/System/Collections/Concurrent/ConcurrentDictionary.cs

```diff
@@ -2132,6 +2124,42 @@ private ReadOnlyCollection<TValue> GetValues()
             }
         }
 
+        /// <summary>Whether T's type can be written atomically (i.e., with no danger of torn reads).</summary>
+        private static bool IsWriteAtomic<T>()
+        {
+            // Section 12.6.6 of ECMA CLI explains which types can be read and written atomically without
+            // the risk of tearing. See https://www.ecma-international.org/publications/files/ECMA-ST/ECMA-335.pdf
+
+            if (!typeof(T).IsValueType ||
+                typeof(T) == typeof(IntPtr) ||
+                typeof(T) == typeof(UIntPtr))
+            {
+                return true;
+            }
```

Comment:
Should this use `typeof` checks for everything instead of a mix of `typeof` checks and `GetTypeCode`?

Reply:
If you would rather prefer to omit the ConcurrentDictionary change from this PR, it would be ok with me.

Reply:
I've reverted the ConcurrentDictionary and Json changes for now.